### PR TITLE
feat: honor displayName of context types

### PR DESCRIPTION
**Pull request by eps1lon**
## Summary

Any component name in errors/warnings that occur during server-side rendering ignore the `displayName`. However, react-devtools does use this property when computing the component name. The [docs mention this property](https://reactjs.org/docs/context.html#contextdisplayname) but are very specific about dev tools usage. Is this limitation intentional and if so: why?

## Test Plan

Tested by triggering a prop-types error.

### CODE REVIEW

**Comment by eps1lon** [comment]
Changed the testing approach and rebase. @bvaughn could you take a look?

File: packages/react/src/__tests__/ReactContext-test.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        </Context.Provider>,
+      );
+    }).toErrorDev(
+      'Warning: Failed prop type: The prop `dontPassToSeeErrorStack` is marked as required in `Validator`, but its value is `undefined`.\n' +
+        '    in Validator (at **)\n' +
+        '    in MyContextType.Consumer (at **)\n' +
+        '    in MyContextType.Provider (at **)',
+    );
+  });
+});
```
Comment (bvaughn) [inline packages/react/src/__tests__/ReactContext-test.js]: nit: This test should probably just be in [`packages/react/src/__tests__/ReactContextValidator-test`](https://github.com/facebook/react/blob/master/packages/react/src/__tests__/ReactContextValidator-test.js) with the other validation tests.

Comment (eps1lon) [reply packages/react/src/__tests__/ReactContext-test.js]: Done.

**Comment by trueadm** [comment]
I've had to revert this PR. Details are in https://github.com/facebook/react/pull/18223.

**Comment by bvaughn** [comment]
Thanks, Dominic!

**Pull request merged**